docs/2024/puzzles/day06.md

```diff
@@ -1,8 +1,214 @@
+  g.simulateWithinLab(startingPov).map(_._1)
+      .foldLeft(g.lab)((lab, coord) => lab.replaceWith(coord._1, coord._2, 'X'))
+```
+
+This function takes all the coordinates the guard visited and replaces the corresponding characters in the lab with an 'X'. This way, we can visualize the lab with the guard's path, just as proposed in the problem statement.
+
+## Part 2
+
+The part 2 asks in how many places we can place a obstacle so that the guard loops forever. To solve this, we will use brute force as it solves it within reasonable time.
+
+A key observation, is that the guard is looping if the path returned by the simulate function is at least longer than the total area of the lab + 1. Indeed, if the guard has visited all locations within the lab more than once without exiting, she is looping.
+She could (and most likely will) visit a subset of the locations, but if the path is at least as long as the total area of the lab + 1, she is looping.
+Please note that we could detect the loop with a smaller path, for example, by checking when we detect a point of view that was already visited. However, this is a simple and elegant solution that works well for the input size.
+
+So we define a `looping` function as follows:
+
+```scala
+def looping(guard: Guard, startingPov: PointOfView): Boolean = 
+  val followedPath = guard.simulateWithinLab(startingPov).take(guard.lab.eastWestLength * guard.lab.northSouthLength + 1)
+  followedPath.size >= guard.lab.eastWestLength * guard.lab.northSouthLength
+````
+
+To make sure the function terminates, we can `take()` with the minimum number of steps we require to detect a loop, which is the total area of the lab + 1. So the returned path will have at most this length. Remember that the `LazyList` is infinite in the case of looping, so if we just check the length of the path, the function will not terminate.
+
+Once we have a function to detect whether the guard loops or not given a lab disposition and starting position, we can write a function that computes how many different obstacle positions lead to a looping guard.
+
+To do so, we will create a list of all possible obstacle positions, which are all positions in the lab except for the starting position, and create copies of the lab with an obstacle in each of these positions.
+We then count how many of those lead to a looping guard:
+
+```scala
+def possibleObstaclesPositionsNumber(g: Guard, startingPov: PointOfView): Int = 
+    val possibleObstaclesPositions = 
+      (
+        for 
+          x <- 0 to g.lab.eastWestLength
+          y <- 0 to g.lab.northSouthLength
+          if g.lab.isWithinLab(x, y) && (x, y) != startingPov._1
+        yield (x, y)
+      )
+
+    val newPossibleGuards= 
+      possibleObstaclesPositions.map(obstaclePos => 
+        val newLab = g.lab.replaceWith(obstaclePos._1, obstaclePos._2, '#')
+        Guard(newLab)
+      )
+
+    val parList = newPossibleGuards.par
+    val forkJoinPool = new ForkJoinPool(8)
```

Comment:
it seems unusual to me to bother configuring the fork join pool; wouldn't the default behavior be fine...?

Reply:
(Even if you feel that it's better that way and don't want to remove it, at minimum I think it should be justified somehow in the article.)

Reply:
Yeah, you're right, makes sense. Let me remove it!
(I was experimenting with larger number of threads while writing the solution, and it is a leftover)